File: 2447-number-of-subarrays-with-gcd-equal-to-k/2447-number-of-subarrays-with-gcd-equal-to-k.cpp
class Solution {
public:
    int subarrayGCD(vector<int>& nums, int k) {
        int res = 0;
    unordered_map<int, int> gcds;
    for (int i = 0; i < nums.size(); ++i) {
        unordered_map<int, int> gcds1;
        if (nums[i] % k == 0) {
            ++gcds[nums[i]];
            for (auto [prev_gcd, cnt] : gcds)
                gcds1[gcd(prev_gcd, nums[i])] += cnt;
        }
        res += gcds1[k];
        swap(gcds, gcds1);
    }
    return res;
    }
};

File: 2447-number-of-subarrays-with-gcd-equal-to-k/2447-number-of-subarrays-with-gcd-equal-to-k.java
class Solution {
     public static int gcd(int a, int b) {
        if(a == 0)
            return b;
         
         return gcd(b%a, a);
    }
    public int subarrayGCD(int[] nums, int k) {
        int n = nums.length;
        int ans = 0;

        for (int i = 0; i < n; i++) {
            int curr_gcd = 0;
            for (int j = i; j < n; j++) {
                curr_gcd = gcd(curr_gcd, nums[j]);
                if(curr_gcd == k)
                    ans++;
            }
        }

        return ans;
    }

   
}